core/network/src/consensus_gossip.rs

```diff
@@ -344,7 +353,19 @@ impl<B: BlockT> ConsensusGossip<B> {
 			let engine_id = message.engine_id;
 			//validate the message
 			if let Some(validator) = self.validators.get(&engine_id) {
-				validator.on_status(who, &message.data)
+				let topics = match validator.on_status(who, &message.data) {
+					StatusResult::Ignore => Vec::new(),
+					StatusResult::BroadcastTopics(topics) => topics,
+				};
+
+				for topic in topics {
+					let messages: Vec<_> = self.messages.iter()
+						.filter_map(|entry| if entry.topic == topic { Some(entry.message.clone()) } else { None })
```

Comment:
i wonder if the loop is faster to do in the other direction

Reply:
> we would want to not loop if topics.is_empty

But that's the case now. The top level loop will not be executed.

Reply:
right, i didn't mean to imply that we didn't do that now. Iterating `self.messages` several times just seems like it might be slower to me. The method i proposed is a bit CPU cache-friendlier

Reply:
My assumption is that most of the times `topics` will be empty indeed, and that would save an iteration on messages.

Reply:
that's why I suggested that, if we did turn the loop inside-out, we would need a special case for `topics.is_empty`